org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/HiddenRegionPartAssociator.java

```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.xtext.formatting2.regionaccess.internal;
+
+import java.util.List;
+import java.util.function.BiConsumer;
+
+import org.eclipse.xtext.formatting2.regionaccess.HiddenRegionPartAssociation;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegion;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegionPart;
+import org.eclipse.xtext.formatting2.regionaccess.IHiddenRegionPartAssociator;
+import org.eclipse.xtext.formatting2.regionaccess.IWhitespace;
+
+/**
+ * @author Moritz Eysholdt - Initial contribution and API
```

Comment:
do i see it correct that the leading comments are previous and trainling comments are next?

```
/*xxxx*/ <-- PREVIOUS
		8
		/*aaaaa*/ <-- next
		a
		b
		c
		// last <-- next
```

Reply:
I'm using the names as follows:
trailing comments are associated with the PREVIOUS semantic region. 
leading comments are associated with the NEXT region. 
```
/*xxxx*/ <-- associated with NEXT semantic region '8'.
8
/*aaaaa*/ <-- associated with NEXT semantic region 'a'.
a
b
c // last <-- associated with PREVIOUS semantic region 'c'.
```

Reply:
but

```
/*xxxx*/
		8
		/*aaaaa*/
		a
		b
		c
		// last
		'''.toTextRegionAccess
```

outputs

```
 0    H "/*xxxx*/"           Comment:TerminalRule'ML_COMMENT' Association:PREVIOUS
    9   "\n"                 Whitespace:TerminalRule'WS'
      B ValueList'[a, b, c]' Root
 9  1  S "8"                  Root:'8'
```

Reply:
That was due to the comments being the first/last element in the file. It kinda made sense for the first comment because often there is a copyright comment at the top of a file which should not be moved along with semantic regions. But I see that "PREVIOUS" and "NEXT" don't really make sense here. I've fixed that and added a test case.